example_updates/nominal_update/example-a-b-update.sh

```diff
@@ -0,0 +1,768 @@
+public_key_file="/adukey/public.pem"
+
+#
+# For install task, the --image-file <image_file_name> option must be specified.
+#
+image_file=""
+
+#
+# Device Update specific arguments
+#
+check_is_installed=
+installed_criteria=
+do_download_action=
+do_install_action=
+do_apply_action=
+do_cancel_action=
+
+restart_to_apply=
+restart_agent_to_apply=
+
+#
+# Remaining aguments and parameters
+#
+PARAMS=
+
+#
+# Output, Logs, and Result helper functions.
+#
+_timestamp=
+
+# SWUpdate doesn't support everything necessary for the dual-copy or A/B update strategy.
+# Here we figure out the current OS partition and then set some environment variables
+# that we use to tell swupdate which partition to target.
+rootfs_dev=$(mount | grep "on / type" | cut -d':' -f 2 | cut -d' ' -f 1)
+if [[ $rootfs_dev == '/dev/mmcblk0p2' ]]; then
+    selection="stable,copy2"
+    current_partition=2
+    update_partition=3
+else
+    selection="stable,copy1"
+    current_partition=3
+    update_partition=2
+fi
+
+update_timestamp() {
+    # See https://man7.org/linux/man-pages/man1/date.1.html
+    _timestamp="$(date +'%Y/%m/%d:%H%M%S')"
+}
+
+log_debug() {
+    if [ $log_level -gt 0 ]; then
+        return
+    fi
+    log "$log_debug_pref" "$@"
+}
+
+log_info() {
+    if [ $log_level -gt 1 ]; then
+        return
+    fi
+    log "$log_info_pref" "$@"
+}
+
+log_warn() {
+    if [ $log_level -gt 2 ]; then
+        return
+    fi
+    log "$log_warn_pref" "$@"
+}
+
+log_error() {
+    if [ $log_level -gt 3 ]; then
+        return
+    fi
+    log "$log_error_pref" "$@"
+}
+
+log() {
+    update_timestamp
+    if [ -z $log_file ]; then
+        echo -e "[$_timestamp]" "$@" >&1
+    else
+        echo "[$_timestamp]" "$@" >> $log_file
+    fi
+}
+
+output() {
+    update_timestamp
+    if [ -z $output_file ]; then
+        echo "[$_timestamp]" "$@" >&1
+    else
+        echo "[$_timestamp]" "$@" >> "$output_file"
+    fi
+}
+
+result() {
+    # NOTE: don't insert timestamp in result file.
+    if [ -z $result_file ]; then
+        echo "$@" >&1
+    else
+        echo "$@" > "$result_file"
+    fi
+}
+
+#
+# Helper function for creating extended result code that indicates
+# errors from this script.
+# Note: these error range (0x30101000 - 0x30101fff) a free to use.
+#
+# Usage: make_swupdate_handler_erc error_value result_variable
+#
+# e.g.
+#         error_code=20
+#         RESULT=0
+#         make_swupdate_handler_erc RESULT $error_code
+#         echo $RESULT
+#
+#  (RESULT is 0x30101014)
+#
+make_swupdate_handler_erc() {
+    local base_erc=0x30101000
+    local -n res=$2 # name reference
+    res=$((base_erc + $1))
+}
+
+# usage: make_aduc_result_json $resultCode $extendedResultCode $resultDetails <out param>
+# shellcheck disable=SC2034
+make_aduc_result_json() {
+    local -n res=$4 # name reference
+    res="{\"resultCode\":$1, \"extendedResultCode\":$2,\"resultDetails\":\"$3\"}"
+}
+
+#
+# Usage
+#
+print_help() {
+    echo ""
```

Comment:
Not necessary to repeat echo and double-quote lines if using a HEREDOC for multiple lines.